src/Flysystem/Plugin/Stat.php

```diff
@@ -139,6 +139,38 @@ protected function getWithMetadata($path, array $ignore)
         return $metadata;
     }
 
+    /**
+     * Normalize a permissions string.
+     *
+     * @param string $permissions
+     *
+     * @return int
+     */
+    protected function normalizePermissions($permissions)
+    {
+      if (is_numeric($permissions)) {
+        return $permissions & 0777;
+      }
+
+      // remove the type identifier
+      $permissions = substr($permissions, 1);
```

Comment:
```suggestion
      $permissions = substr($permissions, -9);
```

This will work with or without the initial `d` or `-` for regular file vs. directory. I think this also makes it clear that we expect 9 characters.

I considered padding with `-` characters to ensure 9 characters, but I am not sure when that would be likely to make a difference. Pad on the left or the right? Maybe it is a bad idea to fix problems that no one has.

Reply:
I feel like the original code is closer to what the comment says it's trying to do, but indeed `-9` is clearer for the _actual_ intent here: extracting the permission groups in isolation.

Padding would make it mildly safer to use with the below, and if you're using `substr` from the right then I'd say it makes sense to pad the left, if at all.

```suggestion
      $permissions = substr(str_pad($permissions, 9, '-', STR_PAD_LEFT), -9);
```

The comment could be updated to:

```php
// Extract the permission groups.
```

The trickiest thing is how vague the file metadata is with Flysystem, but I suppose it has to be.